master/internal/api_resourcepool.go

```diff
@@ -28,3 +36,150 @@ func (a *apiServer) GetResourcePools(
+// for the resource pool state streaming API.
+type ResourcePoolAPIStatesStreamActor struct {
+	resp         apiv1.Determined_ResourcePoolStatesServer
+	tickInterval time.Duration
+	lastStates   []*resourcepoolv1.ResourcePoolState
+}
+
+// NewResourcePoolAPIStatesStreamActor creates a new actor that streams updates until
+// the stream finishes.
+func NewResourcePoolAPIStatesStreamActor(
+	server apiv1.Determined_ResourcePoolStatesServer,
+) *ResourcePoolAPIStatesStreamActor {
+	return &ResourcePoolAPIStatesStreamActor{
+		resp:         server,
+		tickInterval: tickInterval,
+		lastStates:   nil,
+	}
+}
+
+// Receive implements the actor's "receive message and act" loop.
+func (a *ResourcePoolAPIStatesStreamActor) Receive(ctx *actor.Context) error {
+	type tick struct{}
+	switch ctx.Message().(type) {
+	case actor.PreStart:
+		resp := ctx.Ask(sproto.GetCurrentRM(ctx.Self().System()),
+			resourcemanagers.GetResourceSummaries{}).Get()
+		summaries := resp.(resourcemanagers.ResourceSummaries).Summaries
+
+		states := make([]*resourcepoolv1.ResourcePoolState, 0, len(summaries))
+		for _, summary := range summaries {
+			state := resourcepoolv1.ResourcePoolState{
+				Name:                 summary.Name,
+				NumAgents:            int32(summary.NumAgents),
+				SlotsAvailable:       int32(summary.SlotsAvailable),
+				SlotsUsed:            int32(summary.SlotsUsed),
+				CpuContainerCapacity: int32(summary.CPUContainerCapacity),
+				CpuContainersRunning: int32(summary.CPUContainersRunning),
+			}
+			states = append(states, &state)
+		}
+		a.lastStates = states
+
+		err := a.resp.Send(&apiv1.ResourcePoolStatesResponse{
+			States: &resourcepoolv1.ResourcePoolStates{
+				ResourcePools: states,
+			},
+		})
+		if err != nil {
+			return fmt.Errorf("failed while processing batch: %w", err)
+		}
+
+		actors.NotifyAfter(ctx, a.tickInterval, tick{})
+
+	case tick:
+		if a.resp.Context().Err() != nil {
+			ctx.Self().Stop()
+			return nil
+		}
+
+		// Get the full state from the resource pools
+		resp := ctx.Ask(sproto.GetCurrentRM(ctx.Self().System()),
+			resourcemanagers.GetResourceSummaries{}).Get()
+		summaries := resp.(resourcemanagers.ResourceSummaries).Summaries
+
+		currentStates := make([]*resourcepoolv1.ResourcePoolState, 0, len(summaries))
+		for _, summary := range summaries {
+			state := resourcepoolv1.ResourcePoolState{
+				Name:                 summary.Name,
+				NumAgents:            int32(summary.NumAgents),
+				SlotsAvailable:       int32(summary.SlotsAvailable),
+				SlotsUsed:            int32(summary.SlotsUsed),
+				CpuContainerCapacity: int32(summary.CPUContainerCapacity),
+				CpuContainersRunning: int32(summary.CPUContainersRunning),
+			}
+			currentStates = append(currentStates, &state)
+		}
+
+		// Convert full state into diff to send to client
+		updatesToSend := calculateUpdates(a.lastStates, currentStates)
+		a.lastStates = currentStates
+
+		err := a.resp.Send(&apiv1.ResourcePoolStatesResponse{
```

Comment:
suggestion: wdy about only sending updates when there is at least one?

Reply:
Yeah that's a good idea